### PR TITLE
[AP-104] Dashboard visuals complete

### PR DESCRIPTION
**Pull request by AdrienKamran**

### CODE REVIEW

File: backend/controllers/saleController.js [hunk 1]
```diff
@@ -105,4 +105,160 @@ exports.deleteSale = asyncHandler(async (req, res, next) => {
...
+        payload: datasets
+    })
+});
+
+function formatPoints(data, type) {
+    let labels = []
+    let points = []
+    let formattedPoints = [labels, points]
+    let dateString = ""
+    console.log("[BACKEND] Formatting points by "+type)
```
Comment (AbdulSir) [inline backend/controllers/saleController.js:216]: Remove console.log in this file

File: backend/routes/dashboardVisuals.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+const express = require('express');
+const router = express.Router({ mergeParams: true });
+
+const {
+    getSalesByTime
+} = require('../controllers/saleController');
+
+// get authentication middleware
+const { protect, hasPermissions } = require('../middleware/auth');
```
Comment (AbdulSir) [inline backend/routes/dashboardVisuals.js:9]: Remove hasPermissions, it's not being used.

File: backend/routes/dashboardVisuals.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+// add .get(protect, getSalesByTime) later
+router.route('/visual3')
+    .get(getSalesByTime)
```
Comment (AbdulSir) [inline backend/routes/dashboardVisuals.js:13]: Add the protect middileware before the getSalesByTime

File: backend/routes/vehicle.js [hunk 4]
```diff
@@ -28,6 +30,12 @@ router.route('/')
 router.route('/dealership/:dealershipId')
   .get(protect, hasPermissions('View Vehicles'), getVehicles);
 
+router.route('/dealership/:dealershipId/notSold')
+    .get(protect, hasPermissions('View Vehicles'), getNotSoldVehicles);
+
+router.route('/dealership/:dealershipId/stale')
+    .get(getStaleVehicles);
```
Comment (AbdulSir) [inline backend/routes/vehicle.js:37]: Add the protect middleware

File: frontend/src/views/charts/DashboardLineChart.vue [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        url: `${this.$store.state.api}/dashboard-visuals/dealership/${dealership}/visual3`,
+        headers: {
+          Authorization: `Bearer ${this.$store.state.auth.token}`,
+        },
+      })
+        // From this point, to ensure sync, all processing methods MUST reside in this block
+        .then((response) => {
+          console.log("Sales requested successfully!")
+          this.dataPayload = response.data.payload
+          console.log(this.dataPayload)
```
Comment (AbdulSir) [inline frontend/src/views/charts/DashboardLineChart.vue:68]: remove console.log

File: frontend/src/views/charts/DashboardLineChart.vue [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+          this.yearlyLabels = this.dataPayload.formattedSalesByYear[0]
+          this.yearlyPoints = this.dataPayload.formattedSalesByYear[1]
+          this.dataLabels = this.dataPayload.formattedSalesByMonth[0]
+          this.dataPoints[0].data = this.dataPayload.formattedSalesByMonth[1]
+          this.useDataSet(this.timescale, this.periods);
+        })
+        .catch((error) => {
+          console.log(error);
+          // Show an error message instead of showing the 404 page
+          this.$router.replace("/pages/404");
```
Comment (AbdulSir) [inline frontend/src/views/charts/DashboardLineChart.vue:80]: remove this line of code. we don't want to redirect to 404 if one of the visuals is causing errors

File: frontend/src/views/charts/DashboardLineChart.vue [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        else if (period == "Last 12") {
+          this.dataLabels = this.yearlyLabels.slice(-12)
+          this.dataPoints[0].data = this.yearlyPoints.slice(-12)
+        }
+      }
+    },
+  },
+  // Actions to take the instant the component is loaded
+  mounted() {
+    console.log("Dashboard chart mounted -> "+`${this.dealership}`);
```
Comment (AbdulSir) [inline frontend/src/views/charts/DashboardLineChart.vue:112]: remove console.log

**Review by AbdulSir — CHANGES_REQUESTED** [review]
There are many console.log that needs to be removed. Also, use the protect middleware when necessary

File: backend/controllers/saleController.js (hunk 1 above)
Comment (AdrienKamran) [reply backend/controllers/saleController.js]: Fixed!

File: backend/routes/dashboardVisuals.js (hunk 2 above)
Comment (AdrienKamran) [reply backend/routes/dashboardVisuals.js:9]: Fixed!

File: backend/routes/dashboardVisuals.js (hunk 3 above)
Comment (AdrienKamran) [reply backend/routes/dashboardVisuals.js]: Fixed!

File: backend/routes/vehicle.js (hunk 4 above)
Comment (AdrienKamran) [reply backend/routes/vehicle.js]: Fixed!

File: frontend/src/views/charts/DashboardLineChart.vue (hunk 5 above)
Comment (AdrienKamran) [reply frontend/src/views/charts/DashboardLineChart.vue]: Fixed!

File: frontend/src/views/charts/DashboardLineChart.vue (hunk 6 above)
Comment (AdrienKamran) [reply frontend/src/views/charts/DashboardLineChart.vue]: Fixed!

File: frontend/src/views/charts/DashboardLineChart.vue (hunk 7 above)
Comment (AdrienKamran) [reply frontend/src/views/charts/DashboardLineChart.vue]: Fixed!

**Review by AbdulSir — APPROVED** [review]
Code looks better now. Thanks for addressing my comments

**Review by purplesmurf1998 — APPROVED** [review]
Charts work very well and look nice. I'll approve and we can merge once the changes mentioned by @AbdulSir are resolved.

**Pull request merged**